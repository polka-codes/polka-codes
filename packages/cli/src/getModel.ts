import { createAnthropic } from '@ai-sdk/anthropic'
import { createDeepSeek } from '@ai-sdk/deepseek'
import { createOpenAI } from '@ai-sdk/openai'
import type { LanguageModelV2 } from '@ai-sdk/provider'
import { createOpenRouter } from '@openrouter/ai-sdk-provider'
import { createOllama } from 'ollama-ai-provider-v2'

export enum AiProvider {
  Anthropic = 'anthropic',
  Ollama = 'ollama',
  DeepSeek = 'deepseek',
  OpenRouter = 'openrouter',
  OpenAI = 'openai',
}

export type ModelConfig = {
  provider: AiProvider
  model: string
  apiKey?: string
  baseUrl?: string
}

export const getModel = (config: ModelConfig, debugLogging = false): LanguageModelV2 => {
  const fetchOverride: typeof fetch = debugLogging
    ? ((async (url, options) => {
        // log outgoing request
        console.log('-> Request URL:', url)
        console.log('-> Request Headers:', options?.headers)
        console.log('-> Request Body:', JSON.parse(options?.body as any))

        const res = await fetch(url, options)
        console.log('<- Response Status:', res.status)

        const contentType = res.headers.get('content-type') || ''
        // if it's a streaming response, clone the body stream
        if (contentType.includes('text/event-stream') && res.body) {
          const [branch, clientStream] = res.body.tee()
          // consume branch to log chunks
          ;(async () => {
            const reader = branch.getReader()
            const decoder = new TextDecoder()
            let done = false
            while (!done) {
              const { value, done: d } = await reader.read()
              done = d
              if (value) {
                const text = decoder.decode(value)
                console.log('<- Stream chunk:', text.replace(/\n/g, '\\n'))
              }
            }
          })()
          // return the other branch to the SDK
          return new Response(clientStream, {
            headers: res.headers,
            status: res.status,
          })
        }

        // non-stream: read and log full body
        const full = await res.text()
        console.log('<- Response Body:')
        console.dir(JSON.parse(full), { depth: null })
        return new Response(full, {
          headers: res.headers,
          status: res.status,
        })
      }) as typeof fetch)
    : fetch

  switch (config.provider) {
    case AiProvider.Anthropic: {
      const anthropic = createAnthropic({
        apiKey: config.apiKey,
        baseURL: config.baseUrl,
        fetch: fetchOverride,
      })
      return anthropic(config.model)
    }

    case AiProvider.Ollama: {
      const ollama = createOllama({
        baseURL: config.baseUrl,
        fetch: fetchOverride,
      })
      return ollama(config.model)
    }

    case AiProvider.DeepSeek: {
      const deepseek = createDeepSeek({
        apiKey: config.apiKey,
        baseURL: config.baseUrl,
        fetch: fetchOverride,
      })
      return deepseek(config.model)
    }

    case AiProvider.OpenRouter: {
      const openrouter = createOpenRouter({
        apiKey: config.apiKey,
        baseURL: config.baseUrl,
        fetch: fetchOverride,
      })
      return openrouter.chat(config.model, {
        usage: { include: true },
      })
    }

    case AiProvider.OpenAI: {
      const openai = createOpenAI({
        apiKey: config.apiKey,
        baseURL: config.baseUrl,
        fetch: fetchOverride,
      })

      return openai(config.model)
    }
  }
}
