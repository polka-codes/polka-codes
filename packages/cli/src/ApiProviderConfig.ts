import type { Config } from '@polka-codes/cli-shared'
import type { ToolFormat } from '@polka-codes/core'
import { AiProvider } from './getModel'

const isSonnet4 = (model: string) => model.includes('claude-sonnet-4')

const getToolFormat = (model: string, toolFormat?: ToolFormat) => {
  if (toolFormat) {
    return toolFormat
  }
  if (isSonnet4(model)) {
    return 'native'
  }
  return 'polka-codes'
}

const defaultModels = {
  [AiProvider.Anthropic]: 'claude-sonnet-4-20250514',
  [AiProvider.Ollama]: 'deepseek-r1:32b',
  [AiProvider.DeepSeek]: 'deepseek-chat',
  [AiProvider.OpenRouter]: 'google/gemini-2.5-pro',
  [AiProvider.OpenAI]: 'gpt-4.1',
  [AiProvider.GoogleVertex]: 'gemini-2.5-pro',
}

export class ApiProviderConfig {
  readonly defaultProvider?: AiProvider
  readonly providers: Readonly<Partial<Record<AiProvider, { apiKey: string; defaultModel?: string; defaultParameters: any }>>>
  readonly commands?: Partial<Record<string, { provider?: AiProvider; model?: string; parameters: any; toolFormat?: ToolFormat }>>
  readonly agents?: Partial<Record<string, { provider?: AiProvider; model?: string; parameters: any; toolFormat?: ToolFormat }>>
  readonly defaultParameters: any
  readonly toolFormat?: ToolFormat

  constructor(config: Config) {
    this.defaultProvider = config.defaultProvider as AiProvider | undefined
    this.defaultParameters = config.defaultParameters ?? {}
    this.providers = config.providers ?? {}
    this.commands = config.commands as Partial<
      Record<string, { provider?: AiProvider; model?: string; parameters: any; toolFormat?: ToolFormat }>
    >
    this.agents = config.agents as Partial<
      Record<string, { provider?: AiProvider; model?: string; parameters: any; toolFormat?: ToolFormat }>
    >
    this.toolFormat = config.toolFormat
  }

  getConfigForCommand(command: string) {
    // TODO: strong type command
    const { provider, model, parameters, toolFormat } = this.commands?.[command] ?? this.commands?.default ?? {}
    const finalProvider = provider ?? this.defaultProvider
    if (!finalProvider) {
      return undefined
    }
    const { apiKey, defaultModel, defaultParameters, location, project, keyFile } = (this.providers[finalProvider] as any) ?? {}
    const finalModel = model ?? defaultModel ?? defaultModels[finalProvider]
    const finalParameters = {
      ...this.defaultParameters,
      ...(defaultParameters ?? {}),
      ...(parameters ?? {}),
    }
    const finalToolFormat = getToolFormat(finalModel, toolFormat ?? this.toolFormat)
    return {
      provider: finalProvider,
      model: finalModel,
      apiKey,
      location,
      project,
      keyFile,
      parameters: finalParameters,
      toolFormat: finalToolFormat,
    }
  }

  getConfigForAgent(agent: string) {
    // TODO: strong type agent
    const { provider, model, parameters, toolFormat } = this.agents?.[agent] ?? this.agents?.default ?? {}
    const finalProvider = provider ?? this.defaultProvider
    if (!finalProvider) {
      return undefined
    }
    const { apiKey, defaultModel, defaultParameters, location, project, keyFile } = (this.providers[finalProvider] as any) ?? {}
    const finalModel = model ?? defaultModel ?? defaultModels[finalProvider]
    const finalParameters = {
      ...this.defaultParameters,
      ...(defaultParameters ?? {}),
      ...(parameters ?? {}),
    }
    const finalToolFormat = getToolFormat(finalModel, toolFormat ?? this.toolFormat)
    return {
      provider: finalProvider,
      model: finalModel,
      apiKey,
      location,
      project,
      keyFile,
      parameters: finalParameters,
      toolFormat: finalToolFormat,
    }
  }
}
