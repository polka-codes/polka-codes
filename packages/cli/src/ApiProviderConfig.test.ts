// Generated by polka.codes

import { describe, expect, test } from 'bun:test'
import type { Config } from '@polka-codes/cli-shared'
import { ApiProviderConfig } from './ApiProviderConfig'
import { AiProvider } from './getModel'

describe('ApiProviderConfig', () => {
  // Define test configurations
  const basicConfig: Config = {
    defaultProvider: 'anthropic',
    defaultModel: 'claude-3-opus',
    defaultParameters: {
      temperature: 0.7,
      top_p: 0.9,
    },
    providers: {
      anthropic: {
        apiKey: 'test-api-key',
        defaultModel: 'claude-3-haiku',
        defaultParameters: {
          temperature: 0.5,
          max_tokens: 2000,
        },
      },
      openai: {
        apiKey: 'test-openai-key',
        defaultParameters: {
          temperature: 0.8,
          presence_penalty: 0.2,
        },
      },
    },
  }

  test('constructor initializes providers and parameters', () => {
    const apiConfig = new ApiProviderConfig(basicConfig)

    expect(apiConfig.defaultProvider).toBe(AiProvider.Anthropic)
    expect(apiConfig.providers.anthropic?.apiKey).toBe('test-api-key')
    expect(apiConfig.providers.anthropic?.defaultModel).toBe('claude-3-haiku')
    expect(apiConfig.providers.anthropic?.defaultParameters).toEqual({
      temperature: 0.5,
      max_tokens: 2000,
    })
  })

  test('getConfigForCommand resolves parameters correctly', () => {
    const configWithCommands: Config = {
      ...basicConfig,
      commands: {
        default: {
          provider: AiProvider.Anthropic,
          model: 'claude-3-opus',
          parameters: {
            temperature: 0.6,
            max_tokens: 3000,
          },
        },
        task: {
          model: 'claude-3-sonnet',
          parameters: {
            temperature: 0.8,
          },
        },
      },
    }

    const apiConfig = new ApiProviderConfig(configWithCommands)

    // Test default command config
    const defaultConfig = apiConfig.getConfigForCommand('default')
    expect(defaultConfig).toMatchSnapshot()

    // Test specific command that inherits from default provider parameters
    const taskConfig = apiConfig.getConfigForCommand('task')
    expect(taskConfig).toMatchSnapshot()

    // Test command not defined (should use default)
    const undefinedCommandConfig = apiConfig.getConfigForCommand('undefined-command')
    expect(undefinedCommandConfig).toMatchSnapshot()
  })

  test('getConfigForAgent resolves parameters correctly', () => {
    const configWithAgents: Config = {
      ...basicConfig,
      agents: {
        default: {
          provider: AiProvider.Anthropic,
          model: 'claude-3-opus',
          parameters: {
            temperature: 0.6,
            max_tokens: 3000,
          },
        },
        coder: {
          model: 'claude-3-sonnet',
          parameters: {
            temperature: 0.8,
          },
        },
        analyzer: {
          provider: AiProvider.OpenRouter,
          parameters: {
            frequency_penalty: 0.3,
          },
        },
      },
    }

    const apiConfig = new ApiProviderConfig(configWithAgents)

    // Test default agent config
    const defaultConfig = apiConfig.getConfigForAgent('default')
    expect(defaultConfig).toMatchSnapshot()

    // Test specific agent that inherits provider
    const coderConfig = apiConfig.getConfigForAgent('coder')
    expect(coderConfig).toMatchSnapshot()

    // Test agent with different provider
    const analyzerConfig = apiConfig.getConfigForAgent('analyzer')
    expect(analyzerConfig).toMatchSnapshot()

    // Test agent not defined (should use default)
    const undefinedAgentConfig = apiConfig.getConfigForAgent('undefined-agent')
    expect(undefinedAgentConfig).toMatchSnapshot()
  })

  test('parameters inherit correctly from different levels', () => {
    const complexConfig: Config = {
      defaultProvider: AiProvider.Anthropic,
      defaultModel: 'claude-3-opus',
      defaultParameters: {
        temperature: 0.7,
        top_p: 0.9,
        max_tokens: 4000,
      },
      providers: {
        anthropic: {
          apiKey: 'test-api-key',
          defaultParameters: {
            temperature: 0.5,
            presence_penalty: 0.1,
          },
        },
      },
      agents: {
        default: {
          parameters: {
            temperature: 0.6,
          },
        },
        coder: {
          model: 'claude-3-sonnet',
        },
      },
      commands: {
        default: {
          parameters: {
            top_p: 0.8,
          },
        },
        task: {},
      },
    }

    const apiConfig = new ApiProviderConfig(complexConfig)

    // Test inheritance for agent
    const coderConfig = apiConfig.getConfigForAgent('coder')
    expect(coderConfig?.parameters).toMatchSnapshot()

    // Test inheritance for command
    const taskConfig = apiConfig.getConfigForCommand('task')
    expect(taskConfig?.parameters).toMatchSnapshot()
  })

  test('handles missing providers gracefully', () => {
    const minimalConfig: Config = {
      defaultProvider: AiProvider.Anthropic,
    }

    const apiConfig = new ApiProviderConfig(minimalConfig)

    const agentConfig = apiConfig.getConfigForAgent('coder')
    expect(agentConfig?.parameters).toEqual({})

    const commandConfig = apiConfig.getConfigForCommand('task')
    expect(commandConfig?.parameters).toEqual({})
  })

  test('returns undefined when no provider is configured', () => {
    const emptyConfig: Config = {}

    const apiConfig = new ApiProviderConfig(emptyConfig)

    expect(apiConfig.getConfigForAgent('coder')).toBeUndefined()
    expect(apiConfig.getConfigForCommand('task')).toBeUndefined()
  })

  test('handles global default parameters correctly', () => {
    const configWithGlobalDefaults: Config = {
      defaultProvider: AiProvider.OpenRouter,
      defaultModel: 'gpt-4', // Assuming OpenRouter can use gpt-4
      defaultParameters: {
        temperature: 0.2,
        top_p: 0.2,
        frequency_penalty: 0.2,
      },
      providers: {
        openrouter: {
          // Changed from openai to openrouter
          apiKey: 'test-openrouter-key',
          // No defaultParameters here to test global override
        },
        anthropic: {
          apiKey: 'test-anthropic-key',
          defaultModel: 'claude-3-sonnet',
          defaultParameters: {
            temperature: 0.5, // This should override global temperature for Anthropic
            max_tokens: 1000, // This is specific to Anthropic
          },
        },
      },
      agents: {
        default: {
          // Uses OpenRouter by default, should inherit global params
        },
        coder: {
          provider: AiProvider.Anthropic,
          // Uses Anthropic, should use Anthropic's params, overriding global temp
        },
        analyzer: {
          provider: AiProvider.OpenRouter, // Changed from OpenAI
          parameters: {
            temperature: 0.3, // This should override global and provider temperature for this agent
            top_p: 0.3, // This should override global and provider top_p for this agent
          },
        },
      },
      commands: {
        default: {
          // Uses OpenRouter by default, should inherit global params
        },
        task: {
          provider: AiProvider.Anthropic,
          parameters: {
            temperature: 0.6, // This should override global and Anthropic's temperature
            max_tokens: 1500, // This should override Anthropic's max_tokens
          },
        },
      },
    }

    const apiConfig = new ApiProviderConfig(configWithGlobalDefaults)

    // Agent: default (OpenRouter with global defaults)
    const defaultAgentConfig = apiConfig.getConfigForAgent('default')
    expect(defaultAgentConfig?.parameters).toMatchSnapshot('agent default - global params')

    // Agent: coder (Anthropic, merging global and Anthropic defaults)
    const coderAgentConfig = apiConfig.getConfigForAgent('coder')
    expect(coderAgentConfig?.parameters).toMatchSnapshot('agent coder - anthropic params override global')

    // Agent: analyzer (OpenRouter, with agent-specific overrides)
    const analyzerAgentConfig = apiConfig.getConfigForAgent('analyzer')
    expect(analyzerAgentConfig?.parameters).toMatchSnapshot('agent analyzer - agent params override global')

    // Command: default (OpenRouter with global defaults)
    const defaultCommandConfig = apiConfig.getConfigForCommand('default')
    expect(defaultCommandConfig?.parameters).toMatchSnapshot('command default - global params')

    // Command: task (Anthropic, with command-specific overrides)
    const taskCommandConfig = apiConfig.getConfigForCommand('task')
    expect(taskCommandConfig?.parameters).toMatchSnapshot('command task - command params override anthropic and global')

    // Test with only global defaults and no provider specific params
    const configOnlyGlobal: Config = {
      defaultProvider: AiProvider.OpenRouter, // Changed from OpenAI
      defaultModel: 'gpt-3.5-turbo', // Assuming OpenRouter can use this
      defaultParameters: {
        temperature: 0.1,
        max_tokens: 500,
      },
      providers: {
        openrouter: {
          // Changed from openai to openrouter
          apiKey: 'another-openrouter-key',
          // No defaultParameters
        },
      },
      agents: {
        default: {}, // Should use global defaults
      },
    }
    const apiConfigOnlyGlobal = new ApiProviderConfig(configOnlyGlobal)
    const agentOnlyGlobalConfig = apiConfigOnlyGlobal.getConfigForAgent('default')
    expect(agentOnlyGlobalConfig?.parameters).toMatchSnapshot('agent default - only global params')
  })

  test('resolves toolFormat correctly for agents and commands', () => {
    const config: Config = {
      defaultProvider: AiProvider.Anthropic,
      providers: {
        anthropic: {
          apiKey: 'anthropic-key',
          defaultModel: 'claude-sonnet-4-20250514',
        },
        openrouter: {
          apiKey: 'openrouter-key',
          defaultModel: 'anthropic/claude-sonnet-4',
        },
      },
      agents: {
        // Sonnet model should default to native
        default: {
          provider: AiProvider.Anthropic,
        },
        // Sonnet model on OpenRouter should default to native
        coder: {
          provider: AiProvider.OpenRouter,
        },
        // Explicitly set to polka-codes
        analyzer: {
          provider: AiProvider.Anthropic,
          toolFormat: 'polka-codes',
        },
        // Non-sonnet model should default to polka-codes
        architect: {
          provider: AiProvider.Anthropic,
          model: 'claude-3-opus-20240229',
        },
      },
      commands: {
        // Sonnet model should default to native
        default: {
          provider: AiProvider.Anthropic,
        },
        // Sonnet model on OpenRouter should default to native
        task: {
          provider: AiProvider.OpenRouter,
        },
        // Explicitly set to polka-codes
        commit: {
          provider: AiProvider.Anthropic,
          toolFormat: 'polka-codes',
        },
        // Non-sonnet model should default to polka-codes
        pr: {
          provider: AiProvider.OpenRouter,
          model: 'google/gemini-flash-1.5',
        },
      },
    }

    const apiConfig = new ApiProviderConfig(config)

    // Agent tests
    expect(apiConfig.getConfigForAgent('default')?.toolFormat).toBe('native')
    expect(apiConfig.getConfigForAgent('coder')?.toolFormat).toBe('native')
    expect(apiConfig.getConfigForAgent('analyzer')?.toolFormat).toBe('polka-codes')
    expect(apiConfig.getConfigForAgent('architect')?.toolFormat).toBe('polka-codes')

    // Command tests
    expect(apiConfig.getConfigForCommand('default')?.toolFormat).toBe('native')
    expect(apiConfig.getConfigForCommand('task')?.toolFormat).toBe('native')
    expect(apiConfig.getConfigForCommand('commit')?.toolFormat).toBe('polka-codes')
    expect(apiConfig.getConfigForCommand('pr')?.toolFormat).toBe('polka-codes')
  })
})
