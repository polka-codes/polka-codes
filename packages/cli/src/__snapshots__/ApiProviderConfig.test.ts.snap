// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`ApiProviderConfig getConfigForCommand resolves parameters correctly 1`] = `
{
  "apiKey": "test-api-key",
  "model": "claude-3-opus",
  "parameters": {
    "max_tokens": 3000,
    "temperature": 0.6,
    "top_p": 0.9,
  },
  "provider": "anthropic",
  "toolFormat": "polka-codes",
}
`;

exports[`ApiProviderConfig getConfigForCommand resolves parameters correctly 2`] = `
{
  "apiKey": "test-api-key",
  "model": "claude-3-sonnet",
  "parameters": {
    "max_tokens": 2000,
    "temperature": 0.8,
    "top_p": 0.9,
  },
  "provider": "anthropic",
  "toolFormat": "polka-codes",
}
`;

exports[`ApiProviderConfig getConfigForCommand resolves parameters correctly 3`] = `
{
  "apiKey": "test-api-key",
  "model": "claude-3-opus",
  "parameters": {
    "max_tokens": 3000,
    "temperature": 0.6,
    "top_p": 0.9,
  },
  "provider": "anthropic",
  "toolFormat": "polka-codes",
}
`;

exports[`ApiProviderConfig getConfigForAgent resolves parameters correctly 1`] = `
{
  "apiKey": "test-api-key",
  "model": "claude-3-opus",
  "parameters": {
    "max_tokens": 3000,
    "temperature": 0.6,
    "top_p": 0.9,
  },
  "provider": "anthropic",
  "toolFormat": "polka-codes",
}
`;

exports[`ApiProviderConfig getConfigForAgent resolves parameters correctly 2`] = `
{
  "apiKey": "test-api-key",
  "model": "claude-3-sonnet",
  "parameters": {
    "max_tokens": 2000,
    "temperature": 0.8,
    "top_p": 0.9,
  },
  "provider": "anthropic",
  "toolFormat": "polka-codes",
}
`;

exports[`ApiProviderConfig getConfigForAgent resolves parameters correctly 3`] = `
{
  "apiKey": undefined,
  "model": "google/gemini-2.5-pro",
  "parameters": {
    "frequency_penalty": 0.3,
    "temperature": 0.7,
    "top_p": 0.9,
  },
  "provider": "openrouter",
  "toolFormat": "polka-codes",
}
`;

exports[`ApiProviderConfig getConfigForAgent resolves parameters correctly 4`] = `
{
  "apiKey": "test-api-key",
  "model": "claude-3-opus",
  "parameters": {
    "max_tokens": 3000,
    "temperature": 0.6,
    "top_p": 0.9,
  },
  "provider": "anthropic",
  "toolFormat": "polka-codes",
}
`;

exports[`ApiProviderConfig parameters inherit correctly from different levels 1`] = `
{
  "max_tokens": 4000,
  "presence_penalty": 0.1,
  "temperature": 0.5,
  "top_p": 0.9,
}
`;

exports[`ApiProviderConfig parameters inherit correctly from different levels 2`] = `
{
  "max_tokens": 4000,
  "presence_penalty": 0.1,
  "temperature": 0.5,
  "top_p": 0.9,
}
`;

exports[`ApiProviderConfig handles global default parameters correctly: agent default - global params 1`] = `
{
  "frequency_penalty": 0.2,
  "temperature": 0.2,
  "top_p": 0.2,
}
`;

exports[`ApiProviderConfig handles global default parameters correctly: agent coder - anthropic params override global 1`] = `
{
  "frequency_penalty": 0.2,
  "max_tokens": 1000,
  "temperature": 0.5,
  "top_p": 0.2,
}
`;

exports[`ApiProviderConfig handles global default parameters correctly: agent analyzer - agent params override global 1`] = `
{
  "frequency_penalty": 0.2,
  "temperature": 0.3,
  "top_p": 0.3,
}
`;

exports[`ApiProviderConfig handles global default parameters correctly: command default - global params 1`] = `
{
  "frequency_penalty": 0.2,
  "temperature": 0.2,
  "top_p": 0.2,
}
`;

exports[`ApiProviderConfig handles global default parameters correctly: command task - command params override anthropic and global 1`] = `
{
  "frequency_penalty": 0.2,
  "max_tokens": 1500,
  "temperature": 0.6,
  "top_p": 0.2,
}
`;

exports[`ApiProviderConfig handles global default parameters correctly: agent default - only global params 1`] = `
{
  "max_tokens": 500,
  "temperature": 0.1,
}
`;
